20240808-custom-limit-email-notifications.md

```diff
@@ -0,0 +1,194 @@
+
+The purpose of "Project Limit Notification Emails" is to inform users automatically when they are getting close to, or exceeding their custom limits.
+
+### Goals
+
+1. Give users advance notice that they are getting close to their custom limits (80% used)
+2. Inform users when they have exceeded their custom limits (100% used)
+
+### Approach / Design
+
+Things we want to avoid:
+* unintentionally sending multiple emails per "threshold event" (e.g. storage usage on project abc has exceeded 80% of the custom limit)
+* failing to send an email for a "threshold event" (e.g. satellite restart at the same time as threshold event, resulting in dropped communications)
+
+These constraints require that we make some DB updates.
+
+#### Config Updates
+
+A single feature flag should be added on the satellite to enable "limit threshold email notifications". It should default to be turned off, until it is turned on in production.
+
+#### DB Updates
+
+For each project, we need to track whether emails were sent for each "threshold event":
+
+* storage usage exceeded 80% of custom storage limit
+* storage usage exceeded 100% of custom storage limit
+* bandwidth usage exceeded 80% of custom bandwidth limit
+* bandwidth usage exceeded 100% of custom bandwidth limit
+
+We could add a new column to the projects table for each of these, but we can also minimize the schema change by using the bits of a single int field for each one. Example of one approach accomplishing this via a Go struct:
+
+```golang
+type ProjectNotificationFlags struct {
+	StorageExceeded80  bool
+	StorageExceeded100 bool
+	EgressExceeded80   bool
+	EgressExceeded100  bool
+}
+
+const (
+	storage80Filter  int = 1      // 00...0001
+	storage100Filter int = 1 << 1 // 00...0010
+	egress80Filter   int = 1 << 2 // 00...0100
+	egress100Filter  int = 1 << 3 // 00...1000
+  /*
+    example db values:
+    3 = b11 = storage 80 and storage 100 are true
+    5 = b101 = storage 80 and egress 80 are true
+    15 = b1111 = all four are true
+  */
+)
+
+func ProjectNotificationFlagsFromInt(dbVal int) *ProjectNotificationFlags {
+	p := &ProjectNotificationFlags{}
+	if storage80Filter&dbVal > 0 {
+		p.StorageExceeded80 = true
+	}
+	if storage100Filter&dbVal > 0 {
+		p.StorageExceeded100 = true
+	}
+    ...
+	return p
+}
+
+func (p *ProjectNotificationFlags) Int() int {
+	toReturn := 0
+	if p.StorageExceeded80 {
+		toReturn |= storage80Filter
+	}
+	if p.StorageExceeded100 {
+		toReturn |= storage100Filter
+	}
+    ...
+	return toReturn
+}
+```
+
+If we use this approach, we only need to add a single updatable int column to the projects table, and since int has 32 bits, we would also be able to support any other similar flags we might want in the future, without schema changes.
+
+#### Event detection and email sending in a chore
```

Comment:
TODO: estimate how long it would take to iterate over and check usage for all projects in production - i.e. the maximum delay for email notifications after an event occurs